src/middlewared/middlewared/plugins/vm/devices/usb.py

```diff
@@ -4,48 +4,111 @@
+
     def identity(self):
-        return self.usb_device
+        return self.usb_device or f'{self.usb_details["product_id"]}--{self.usb_details["vendor_id"]}'
 
     def get_vms_using_device(self):
+        if self.usb_device:
+            device_filter = ['attributes.device', '=', self.usb_device]
+        else:
+            device_filter = ['attributes.usb', '=', self.usb_details]
         devs = self.middleware.call_sync(
-            'vm.device.query', [['attributes.device', '=', self.usb_device], ['dtype', '=', 'USB']]
+            'vm.device.query', [device_filter, ['dtype', '=', 'USB']]
         )
         return self.middleware.call_sync('vm.query', [['id', 'in', [dev['vm'] for dev in devs]]])
 
     def get_details(self):
-        return self.middleware.call_sync('vm.device.usb_passthrough_device', self.usb_device)
+        usb_device = self.usb_device
+        if not usb_device and self.usb_details:
+            usb_device = self.middleware.call_sync('vm.device.get_usb_port_from_usb_details', self.usb_details)
+        if usb_device:
+            return self.middleware.call_sync('vm.device.usb_passthrough_device', usb_device)
+        else:
+            return self.middleware.call_sync('vm.device.usb_passthrough_device', str(usb_device))
 
     def is_available(self):
         return self.get_details()['available']
 
     def xml_linux(self, *args, **kwargs):
+        controller_mapping = kwargs.pop('controller_mapping')
         details = self.get_details()['capability']
-        device_xml = create_element(
+        return create_element(
             'hostdev', mode='subsystem', type='usb', managed='yes', attribute_dict={
                 'children': [
                     create_element('source', attribute_dict={'children': [
                         create_element('vendor', id=details['vendor_id']),
                         create_element('product', id=details['product_id']),
                         create_element('address', bus=details['bus'], device=details['device']),
                     ]}),
+                    create_element('address', type='usb', bus=str(controller_mapping[self.controller_type])),
                 ]
             }
         )
-        return device_xml
 
     def _validate(self, device, verrors, old=None, vm_instance=None, update=True):
+        if device['attributes']['device'] and device['attributes']['usb']:
+            verrors.add(
+                'attributes.usb',
+                'Either device must be specified or USB details but not both'
+            )
+
+        if verrors:
+            return
+
+        if device['attributes']['device']:
+            self._validate_usb_port(device, verrors)
+        else:
+            self._validate_usb_details(device, verrors)
+
+    def _validate_usb_details(self, device, verrors):
+        usb_details = device['attributes']['usb']
+        for k in filter(lambda k: not usb_details.get(k), ('product_id', 'vendor_id')):
```

Comment:
What if they are not provided ?

Reply:
Actually i think i should be probably marking this as required there and removing these as `usb` can be null and will be null by default - so this would be redundant then and handled at schema already. Thank you for pointing that out ;)